pkg/storegateway/bucket_chunk_reader.go

```diff
@@ -101,7 +101,7 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, res []seriesEntry, s
 	fetchBegin := time.Now()
 
 	// Get a reader for the required range.
-	reader, err := r.block.chunkRangeReader(ctx, seq, int64(part.Start), int64(part.End-part.Start))
+	reader, err := r.block.chunkRangeReader(ctx, seq, int64(part.Start), int64(part.End-part.Start), chunksPool)
```

Comment:
The caching bucket is configured to cache 16K contiguous subranges of bytes. The size of each slab is 16384 bytes, so it's super close. 

https://github.com/grafana/mimir/blob/0a5e0e33a4bdcfe521e6638643876987b15b219a/pkg/storegateway/series_chunks.go#L26-L27
https://github.com/grafana/mimir/blob/f14724e26e2a9e87a6ac1e5efe2062931d5d5eba/pkg/storegateway/bucket.go#L859
https://github.com/grafana/mimir/blob/4ee079a33a6057e5431d2202b053b6ef8780b9ce/pkg/storage/tsdb/caching_config.go#L45

The caching bucket will effectively reach to the pool for every subrange, but it will also leave some ~2% (384/16384) bytes unused. I wonder if we should have some sort of validation on the subrange config option to make sure it's not more than the slab size. Otherwise the pooling is effectively useless, because we'd allocate a new byte slice on each call ([code](https://github.com/grafana/mimir/blob/f14724e26e2a9e87a6ac1e5efe2062931d5d5eba/pkg/util/pool/pool.go#L182-L184)).

Reply:
I think @pracucci mentioned that he didn't even think the subrange should be possible to configure. That would remove the possibility of changing the configuration to something that would cause all requests to allocate.

Yeah, it does feel like the slab portion of this isn't super useful since they're nearly 1:1 with ranges we're caching. The memcached client _does_ need 2 extra bytes for `CRLF`  so I wouldn't want to make the slab size and sub range size extactly the same. Maybe it would be sufficient to make the slab size a few times bigger than it is today (16384 * 4 or so) since this `SafeSlabPool` is the same one used for chunks fetched object storage.

Reply:
I've gone ahead and changed the slab size from 16K to 64K. This appears to result in slightly fewer allocations but it's difficult to tell based on the noise in the dev environment.

Reply:
i was trying to see what a good slab size is for the other chunks cache PR (https://github.com/grafana/mimir/pull/3968). There each item is in the order of 16K, but could be larger or smaller. I tried with slab sizes of 16 (effectively no pooling), 16K, 1.6M, and 16M

1.6M seemed to give the best results in terms of rate of allocations. And 16 was best for in-use heap.

I wanted a higher number for the slab size because the items aren't aligned to 16K, so it's easier to lose some of the slabs if you only fit fewer items in a slab. Maybe for this caching since all items are exactly 16K you don't need to fit hundreds of items per slab to utilize it. Anyways, this is what i saw
<img width="1673" alt="Screenshot 2023-01-27 at 09 39 58" src="https://user-images.githubusercontent.com/21020035/215045953-30119822-2fa3-496f-a6c7-a60fb5bc1293.png">


<img width="1668" alt="Screenshot 2023-01-27 at 09 50 01" src="https://user-images.githubusercontent.com/21020035/215045946-b424d607-a766-4846-addd-924ee0e22090.png">

Reply:
I can experiment some more today. Besides the slab size, is there anything else you'd like to see in this PR?

Reply:
> Deprecating the subrange size option

LGTM. I would do in a separate PR tho.

Reply:
> i was trying to see what a good slab size is for the other chunks cache PR (https://github.com/grafana/mimir/pull/3968). There each item is in the order of 16K, but could be larger or smaller. I tried with slab sizes of 16 (effectively no pooling), 16K, 1.6M, and 16M

I don't think the numbers you showed justify moving away from 16KB.

Reply:
> I don't think the numbers you showed justify moving away from 16KB. At least, not to 1.6MB which may perform poorly on small requests (because the same slab can't be shared between 2 requests).

Alright. I'll stick with 64K to allow a few cached chunks to be in the same slab without monopolizing large slices of memory from the pool.